sdk/iot/core/inc/az_iot_hub_client.h

```diff
@@ -0,0 +1,505 @@
+    az_iot_hub_client const* client,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    az_span mqtt_password,
+    az_span* out_mqtt_password);
+
+/**
+ *
+ * Properties APIs
+ *
+ *   IoT Hub message properties are used for Device to Cloud (D2C) as well as Cloud to Device (C2D).
+ *   Properties are always appended to the MQTT topic of the published or received message and
+ *   must contain Uri-encoded keys and values.
+ */
+
+/**
+ * @brief Telemetry or C2D properties.
+ *
+ */
+typedef struct az_iot_hub_client_properties
+{
+  struct
+  {
+    az_span properties;
+    uint8_t* current_property;
+  } _internal;
+} az_iot_hub_client_properties;
+
+/**
+ * @brief Initializes the Telemetry or C2D properties.
+ *
+ * @param properties
+ * @param buffer Can either be an empty #az_span or an #az_span containing properly formatted
+ *               properties starting with the '?' character.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result
+az_iot_hub_client_properties_init(az_iot_hub_client_properties* properties, az_span buffer);
+
+/**
+ * @brief Appends a key-value property to the list of properties.
+ *
+ * @param properties
+ * @param name The name of the property.
+ * @param value The value of the property.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_properties_append(
+    az_iot_hub_client_properties* properties,
+    az_span name,
+    az_span value);
+
+/**
+ * @brief Finds the value of a property.
+ * @note This will return the first value of the property with the given name if multiple properties
+ *       with the same key exist.
+ *
+ * @param properties
+ * @param name The name of the property.
+ * @param out_value An #az_span containing the value of the property.
+ * @return #az_result.
+ */
+AZ_NODISCARD az_result az_iot_hub_client_properties_find(
+    az_iot_hub_client_properties* properties,
+    az_span name,
+    az_span* out_value);
+
+/**
+ * @brief Iterates over the list of properties.
+ *
+ * @param properties
+ * @param out An #az_pair containing the key and the value of the next property.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result
+az_iot_hub_client_properties_next(az_iot_hub_client_properties* properties, az_pair* out);
+
+/**
+ *
+ * Telemetry APIs
+ *
+ */
+
+/**
+ * @brief Gets the MQTT topic that must be used for device to cloud telemetry messages.
+ * @note Telemetry MQTT Publish messages must have QoS At Least Once (1).
+ *
+ * @param client
+ * @param properties An optional properties object (can be NULL).
+ * @param mqtt_topic An empty #az_span with sufficient capacity to hold the MQTT topic.
+ * @param out_mqtt_topic The output #az_span containing the MQTT topic.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_telemetry_publish_topic_get(
+    az_iot_hub_client const* client,
+    az_iot_hub_client_properties const* properties,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic);
+
+/**
+ *
+ * Cloud to device (C2D) APIs

```

Comment:
>Cloud to device (C2D) APIs [](start = 3, length = 26)

Should all of this live in one header file?  *In theory* the linker optimizer will be smart enough to cleanup up.  But if stuff is in separate .c/.h makes both MS and customer be more thoughtful RE what they're dragging it at any stage?

Reply:
I think a lot of the original plans for v2 were shooting for single headers. uLib even started making a monolithic `<thing>_api.h` header file which has in it everything the user needs to get going. At that point they can choose to use whatever they want. Actually with that: any ideas on whether we should change the name to `az_iot_hub_api.h`? And provisioning to `az_iot_provisioning_api.h`? Pardon me for the continued name discussions. Feel free to dismiss if its too much

Reply:
@jspaith We've already discussed and decided offline on only 2 header files. This is in-line with AzRTOS as well as other RTOSes. @tawalke was actually debating if we should have only a single .h file for all provided services.

@danewalton we've also settled on having "client" within the name. I'll let Az SDKs comment if we should change.